nx/lib/nx/defn/expr.ex

```diff
@@ -728,7 +728,9 @@ defmodule Nx.Defn.Expr do
       raise "argsort comparator must return a predicate type, got: #{inspect(fun.type)}"
     end
 
-    expr(out, context, :argsort, [tensor, opts, fun])
+    opts = put_in(opts[:comparator_fun], fun)
```

Comment:
I think we should call it :comparator (which is the same name as the input to this function).

Reply:
So I thought, but there is already `:comparator` in the options and BinaryBackend uses it:
https://github.com/elixir-nx/nx/blob/bf0293b7495c5b43d86af94e82ba794deb9acba5/nx/lib/nx/binary_backend.ex#L1709

Reply:
I see. So what if we don't send comparator_fun and we let EXLA normalize :asc and :desc like the binarybackend?

Reply:
What about these validations?

https://github.com/elixir-nx/nx/blob/bf0293b7495c5b43d86af94e82ba794deb9acba5/nx/lib/nx/defn/expr.ex#L723-L729

Reply:
I see... so I think EXLA actually got it right. We don't want to hide the `fun`, which is a Nx.Defn.Expr, inside the options. If we do so, some operations like `Nx.Defn.Tree.traverse_args` won't traverse it. So I think it should be `tensor, opts, comparator`, and leave the original comparison function as `:comparator`. Also please mirror the changes to `sort` too.

Reply:
But the problem is that if we pass the comparator as a separate argument, then this would require `BinaryBackend` to accept such argument, and consequently every backend to do so. I certainly miss something though :>

Reply:
Yes, the backend should receive it and call it unless the comparator function is :asc / :desc.

Reply:
Does this mean we have to build this `Nx.Defn.Expr` comparator on the highest level in `Nx.argsort`?

Reply:
No, Nx.Defn.Expr already builds it for argsort. And then, when evaluating, we unwrap the Nx.Defn.Expr for the function into a regular function. The important is to keep the anonymous function at the top level of the args, so the existing code can see it. If you do that, then you don't need to change anything else. :)

Reply:
My point is that if we want to pass `fun` as an argument and not an option, we would need to change `BinaryBackend.argsort/3` to `BinaryBackend.argsort/4`, which doesn't comply with the backend behaviour, isn't that the case?

Reply:
Yes, we should change the backend behaviour for sort and argsort. :)

Reply:
Thanks! I *think* I got this now, let me know if the updated version makes sense :cat: